app/models/sipity.rb

```diff
@@ -81,6 +83,10 @@ def WorkflowState(input, workflow, &block) # rubocop:disable Naming/MethodName
   end
   module_function :WorkflowState
 
+  ##
+  # A parent error class for all workflow errors caused by bad state
+  class StateError < RuntimeError; end
```

Comment:
@jeremyf i was wondering if you had broader thoughts about sipity error handling/transactionality.

one thing that kept popping up as i was debugging proposed changes is that failed Sipity operations (admittedly, they normally failed because i broke something important) are rarely retry-able. a second attempt will fail as violating a uniqueness constraint for some sipity DB row. the errors raised are pretty much always some kind of ActiveRecord error.

the `StateError` changes included here were what i needed to get reasonably smooth handling in this PR, but it might be nice to do a once over on sipity error handling in general.

also, is it reasonable to try to wrap high-level operations (e.g. `::Hyrax::Workflow::WorkflowFactory.create`) in an ActiveRecord transaction?

Reply:
> Is it reasonable to try to wrap high-level operations (e.g. ::Hyrax::Workflow::WorkflowFactory.create) in an ActiveRecord transaction?

Yes!

Reply:
> one thing that kept popping up as i was debugging proposed changes is that failed Sipity operations (admittedly, they normally failed because i broke something important) are rarely retry-able. a second attempt will fail as violating a uniqueness constraint for some sipity DB row. the errors raised are pretty much always some kind of ActiveRecord error.

> the StateError changes included here were what i needed to get reasonably smooth handling in this PR, but it might be nice to do a once over on sipity error handling in general.

I think much of the errors come from a traditional 🤷 treatment of transactions.

A once over of error handling would be reasonable.  I'm at the over-promised stage of things.  I believe I still owe a read through/documentation of the Sipity workflows.